src/python_tools/tftp_transfer/client.py

```diff
@@ -49,50 +29,85 @@ def main():
         "--file",
         required=True,
         type=pathlib.Path,
-        help="path to the file to be transferred",
+        help="path to the file to be transferred. This should be an absolute path",
+    )
+    ap.add_argument(
+        "--tftp_port",
+        type=int,
+        help="the port to be used for tftp transfers",
+        required=True,
     )
-
-    # TODO (#2237): when announcements are working, we can just obtain the port and ip address from them and the
-    #               following two arguments aren't necessary
-    ap.add_argument("-p", "--port", type=int, help="the port to be used", required=True)
     ap.add_argument(
-        "-ip",
-        "--ip_address",
+        "--announce_port",
+        type=int,
+        help="the port to be used for listening to announcements",
         required=True,
-        type=str,
-        help="ip address of outbound connection",
+    )
+    ap.add_argument(
+        "-r",
+        "--retries",
+        type=int,
+        help="number of times to retry",
+        default=MAX_UPLOAD_RETRIES,
+        required=False,
     )
 
     args = vars(ap.parse_args())
+
     file_path = args["file"]
-    file_name = os.path.basename(args["file"])
+    file_name = os.path.basename(file_path)
     transfer_file_hash = encode_sha256_checksum(file_path)
-    port = args["port"]
-    ip_address = args["ip_address"]
-    client = tftpy.TftpClient(ip_address, port)
-
-    # send file to the robot
-    client.upload(file_name, file_path)
-    while True:
-        # verify the checksum
-        print("Verifying checksum...")
-        # TODO (#2237): currently, the checksum is verified by redownloading the file that we just
-        #       transferred from the robot to a temp directory & making sure the checksum
-        #       of the redownloaded file is the same. When we have announcements, the
-        #       checksum will be reported by the robot and we just compare the hash from the
-        #       announcement
-        with tempfile.TemporaryDirectory() as tmpdirname:
-            transferred_file_path = os.path.join(tmpdirname, file_name)
-            client.download(file_name, transferred_file_path)
-
-            if verify_checksum(transferred_file_path, transfer_file_hash):
+
+    tftp_port = args["tftp_port"]
+    announce_port = args["announce_port"]
+
+    max_retries = args["retries"]
+
+    retries_count = 0
+    # keep track of announcements from robots that have the correct sha256 checksum so that we dont re-upload files
+    # Note: we use a map of robot_id to announcement because announcements are not hashable
+    verified_announcements = {}
+
+    # additional + 1 is due to needing to verify announcements
+    while retries_count < max_retries + 1:
+        announcements = {
+            announcement.robot_id: announcement
+            for announcement in receive_announcements(announce_port)
```

Comment:
I think we should pass in the timeout to `receive_announcements` rather than rely on the default value defined in another file (actually I don't think there should be a default value for this function)